### ISSUE TITLE
copy.deepcopy seems to break SeawatModel because of DataFrame inheritance

### ISSUE THREAD

**Issue by Manangka** [Deltares/imod-python#63]
In GitLab by @Huite on May 14, 2019, 15:24

I'm in the process of updating a test. To avoid side-effects, I'm trying to `copy.deepcopy` the model object.
This appears to be causing issues for the Wel package (which is a subclassed DataFrame).

```python
AttributeError: 'DataFrame' object has no attribute '_pkg_id'
```

I've noticed that `xarray.Dataset` seems to behave very nicely: applying xarray methods (like `.where`) results in the same object being returned. This is not true for subclassed pandas DataFrames. `.loc` returns on a `Wel` object returns a `DataFrame`.

I think it's worth considering whether we should also turn the `Wel` into a subclassed `Dataset`. The only issue is that we're using `pandas` functions to write to ipf. This is likely not a real issue: we can simply call `.to_dataframe()` before writing and use multi-indexes on the Dataset so it resembles a pandas DataFrame.

I don't think any significant changes are necessary for the `Wel` constructor, as it doesn't take a whole DataFrame anyway (just series of numpy arrays).

**Comment by Manangka** [comment]
In GitLab by @visr on Jun 4, 2019, 14:46

If we would change this from pandas to xarray, would you also want to return xarray objects by default when reading IPF files?

For consistency this may be nice, but of course pandas is more stable and has more features and in general is a better fit for tabular data.

**Comment by Manangka** [comment]
In GitLab by @Huite on Jun 9, 2019, 01:29

I think I'd just be pragmatic about it and use a Dataset behind the scenes, that is, within the model. It doesn't matter during the `__init__`: xarray will accept `pandas.Series` objects, which have to given separately anyway (x, y, rate). 

Writing then proceeds with `.to_dataframe` as sketched above. The packages themselves aren't really designed to be objects that a user deals with a lot anyway; they're providing a mapping to MODFLOW terms first and foremost. And if you want to work on it, just call .to_dataframe. You can't reinsert it in the current form anway, since it'll no longer be a `Well` package, but simply a DataFrame!

From a pragmatic point of view, Datasets are just easier in the `model` context. Outside of that context, everything's just a DataFrame. We could think of something clever where it gets converted to a dataframe when you try to do something to it, but I think we'd probably be solving a non-existent problem... my idea is that we simply do it and see if anybody notices (and then complains).

**Comment by Manangka** [comment]
In GitLab by @Huite on Jul 30, 2019, 09:29

Final stuff addressed in: e186e729c3c24c5762d434bc16e3d4963cc5e178

Should be working alright now.

**Comment by Manangka** [comment]
In GitLab by @Huite on Jul 30, 2019, 09:30

closed
